File: cs-core/extra/dsa-js-practice/sortingAlgos/mergeSort.ts
function merge(left: number[], right: number[]): number[] {
  const result: number[] = [];

  let i = 0;
  let j = 0;

  while (i < left.length && j < right.length) {
    const a = left[i];
    const b = right[j];

    if (a < b) {
      result.push(a);
      i++;
    } else {
      result.push(b);
      j++;
    }
  }

  if (i < left.length) {
    for (let k = i; k < left.length; k++) {
      const current = left[k];
      result.push(current);
    }
  }

  if (j < right.length) {
    for (let k = j; k < right.length; k++) {
      const current = right[k];
      result.push(current);
    }
  }

  return result;
}

function mergeSort(list: number[]): number[] {
  if (list.length <= 1) return list;

  const midIndex = Math.floor(list.length / 2);

  const left = mergeSort(list.slice(0, midIndex));
  const right = mergeSort(list.slice(midIndex));

  return merge(left, right);
}

const myArray = [4, 5, 6, 3, 2, 222, 444, 4, 66, 33, 995];

console.log(`
            My array is ${myArray}
            My sorted array is ${mergeSort(myArray)}
            `);

File: cs-core/extra/dsa-python-practice/sorting_algos/merge_sort.py
def merge(left: list[int], right: list[int]) -> list[int]:
    result = []

    i = j = 0

    n, m = len(left), len(right)

    while i < n and j < m:
        a, b = left[i], right[j]

        if a > b:
            result.append(b)
            j += 1
        else:
            result.append(a)
            i += 1

    if i < n:
        result.extend(left[i:])

    if j < m:
        result.extend(right[j:])

    return result


def merge_sort(list: list[int]) -> list[int]:
    n = len(list)

    if n <= 1:
        return list

    mid_index = n // 2

    left = merge_sort(list[:mid_index])
    right = merge_sort(list[mid_index:])

    return merge(left, right)


my_list = [
    1,
    4,
    2,
    5,
    6,
    7,
    8,
]

print(f"""
      My list is {my_list}
      My merged list is: {merge_sort(my_list)}
      """)
